### ISSUE TITLE
Ability to include clj files in resulting jar when AOT'ing a single namespace

### ISSUE THREAD

**Issue by devn** [clojurephant/clojurephant#135]
## Desired Behavior

I have a scenario where I want to AOT a namespace `foo.core`, but would like to keep all non-AOT'd *.clj files in the resulting jar. As it stands, once you add `aotNamespaces = ['foo.core']`, `clj` files get kicked out of the jar.

**Comment by ajoberstar** [comment]
I think the simplest workaround would be something like this:

```groovy
jar {
  from(sourceSets.main.clojure)
}
```

I probably won't add a config option for this unless I get to re-modelling the whole space.

**Comment by jzwolak** [comment]
I'm not sure I understand the use of `aotNamespaces` if non-AOT'd namespaces are _not_ included in the resulting jar. It seems to me that if non-AOT'd namespaces are not going to be included then Clojurephant should be limited to `aotAll` or no AOT. Am I missing something?

I'm having a similar problem where I have a single namespace I wish to AOT and my tests fail because the non-AOT'd namespaces are kicked off the classpath during build. As a result, those namespaces cannot be required by the test namespaces.

As an aside, I appreciate all the work on Clojurephant and I appreciate the difficulty in creating a tool like this.

**Comment by jzwolak** [comment]
The suggest from @ajoberstar on May 24, 2020, works for adding the files to the jar. My use case was slightly different and I want to mention it here because I couldn't use `sourceSets`, AKAIK.

To get the source dirs on the classpath for the tests I used the following:
```
compileTestClojure {
    classpath.from(clojure.builds.main.sourceRoots)
}
```

**Comment by jzwolak** [comment]
The use case that brought this up for me is `gen-class`. I need to use `gen-class` to subclass a Java class. I want to do this in Clojure because my project is currently pure Clojure and it's just a single class with just a tiny bit of functionality that I need to add. I must subclass because it's extending a class, there is unavoidable state involved, and there are methods unique to the subclass. The use of `gen-class` requires AOT, but just for this one Clojure file.

**Comment by ajoberstar** [comment]
@jzwolak My original assumption was that in choosing to AOT a single namespace, that you were indicating the main namespace of an application and thus it was all you needed in the JAR. I haven't used it a lot, so I don't have any personal preference on how it should behave. How do other tools handle this (mainly thinking lein)? If AOT is enabled, does line still JAR up the sources? Does it provide an opt-in/out to include/exclude the source?

**Comment by jzwolak** [comment]
I have created an example Leiningen project using my Retroact source code. My Retroact project uses Gradle as the build system, so there are now two projects - one using Leiningen and one using Gradle. I have compared them side by side and they are both publicly available if you would like to compare them yourself.

Leiningen behaves intuitively as expected. When I set a single namespace as AOT'd, that namespace is compiled before all other tasks are run. Then, the tasks I requested (and their dependencies) are run with the classes from the AOT'd namespace on the classpath. That's it. The jar contains the Clojure source for all my namespaces, including the AOT'd namespace.

When using Clojurephant, nothing works as expected for my use case and I have to add paths and files to 5 different tasks and things are still not working. Well... those 5 tasks work after I update their classpaths and inputs, but many other things still don't work.

In the Leiningen docs it says: "If you have any Java to be compiled in :java-source-paths or Clojure namespaces listed in :aot, they will always be compiled before Leiningen runs any other code, via any run, repl, etc. invocations." which suggests the behavior I witnessed. This quote comes from https://github.com/technomancy/leiningen/blob/stable/doc/TUTORIAL.md

My Retroact projects can be found at:
Gradle based project: https://github.com/jzwolak/retroact (use commit 36743bb to match the Leiningen project)
Leiningen based project: https://github.com/jzwolak/retroact-lein-example

I ran `lein jar` and `jar tf target/retroact-0.1.0-SNAPSHOT.jar` to see the files in the jar. I ran `lein run` to verify the Clojure files can be loaded and run.

Compare that to `./gradlew jar` and `jar tf build/libs/retroact-0.2.1.jar` (ignore the version difference, I didn't set the version in Leiningen).

I find this pretty compelling for changing the way `aotNamespaces` is handled in Clojurephant... though I don't know what the priority is for such changes. I did get the Clojurephant sources and played around with changing things myself... but I do not understand Gradle's model well enough to know what needs changing. I feel like I found the relevant code in Clojurephant, but there's sources, inputs, classpaths, outputs... and multiple tasks... I don't know what goes where and things don't get resolved immediately... so I also don't know what goes "when". :-/

**Comment by jzwolak** [comment]
I'm not sure if Leiningen has an opt in / opt out to include or exclude the source. By default the source is included for all namespaces, including the compiled ones. I could not find anything in a couple minute search of the docs to figure this out.

**Comment by ajoberstar** [comment]
Yep, I'd agree the leiningen's behavior is a good indicator of Clojurephant should behave. I think these are the two spots that would need to change (along with associated tests).

- In [ClojureBasePlugin](https://github.com/clojurephant/clojurephant/blob/main/src/main/java/dev/clojurephant/plugin/clojure/ClojureBasePlugin.java#L73-L79)
- In [ClojureScriptBasePlugin](https://github.com/clojurephant/clojurephant/blob/main/src/main/java/dev/clojurephant/plugin/clojurescript/ClojureScriptBasePlugin.java#L65-L71) -- Would the same apply to ClojureScript?

Presuming we don't see a need to allow someone to opt out of including the source, it can probably change like this:

```java
        if (build.isCompilerConfigured()) {
          return project.files(build.getOutputDir(), build.getSourceRoots());
        } else {
          return build.getSourceRoots();
        }
```

I think we carried the current behavior over from the plugin @cmf wrote, but not sure if he had a strong reason for the behavior or if anyone has started to rely on it since.

**Comment by jzwolak** [comment]
I'm smiling because your proposed code changes are identical (character per character) to what I already tried. They did not work. It's not just that the source must appear in the output directory. The checkClojure task must depend on the compileClojure task - similarly the test tasks need to do this. Also, the exported classpaths for composite builds have to change, too. There are many things that must change.

Since your suggestion is exactly what I started with, I'd be happy to spend a few minutes when I have time to play with this code change and see what I can do and create a better report if I can't figure it out. I have everything setup for testing this use case.

**Comment by jzwolak** [comment]
After a lot of work, I believe a better solution than

```
jar {
  from(sourceSets.main.clojure)
}
```

for including Clojure source in the jar is

```
sourceSets {
    main {
        output.getClassesDirs().from(sourceSets.main.clojure.srcDirs)
    }
}
```

This will not only include the Clojure source in the jar, it will export the source dirs as output dirs for the sourceSet for the main configuration. This second part is crucial for a number of other tasks and features of Gradle, including composite builds. Though the above code may be overriding code from the Clojurephant plugin, I believe this thread has shown that the Clojurephant plugin still needs work in the area of AOT.

**Comment by ajoberstar** [comment]
@jzwolak Good to confirm that approach works for you. That's effectively the same as the change I proposed [above](https://github.com/clojurephant/clojurephant/issues/135#issuecomment-1301443781). I'm curious why that didn't work in your previous testing?

**Comment by jzwolak** [comment]
Gradle doesn't use the jar in composite builds. It uses the output directories of the sourceSet. I analyzed the classpaths using `--debug` on the gradle command line and the classpaths were different when using `jar { .. }` and when using `sourceSets { ... }`. Not only were the classpaths different in the composite builds... there were multiple other tasks in the Clojure project where `sourceSets { ... }` was added that did not need modification but did need modification when using `jar { ... }`.

**Comment by jzwolak** [comment]
Ah, that was my bad for not clicking the link you gave.

I did try the suggestion you made to modify the Clojurephant plugin. It did not work. I'll have to go get more information as to why and post it here.
